src/pip/_internal/vcs/versioncontrol.py

```diff
@@ -232,12 +232,20 @@ def get_backend_for_dir(self, location):
         Return a VersionControl object if a repository of that type is found
         at the given directory.
         """
+        candidates = {}
         for vcs_backend in self._registry.values():
-            if vcs_backend.controls_location(location):
-                logger.debug('Determine that %s uses VCS: %s',
-                             location, vcs_backend.name)
-                return vcs_backend
-        return None
+            root = vcs_backend.get_repository_root(location)
+            if not root:
+                continue
+            logger.debug('Determine that %s uses VCS: %s',
+                         location, vcs_backend.name)
+            candidates[root] = vcs_backend
+
+        # Choose the VCS in the inner-most directory (i.e. path to root
+        # is longest).
+        if not candidates:
+            return None
+        return candidates[max(candidates, key=len)]
```

Comment:
Uh, what?

Reply:
Yes? 😛 

The root would be either cwd or one of its parent directories, so the one longest in length must be the one we’re looking for.

Reply:
I'd suggest moving the variable into a separate line, jsst to make it clearer what's happening. A comment describing why we're using `len` directly, vs something `len(x.split(os.sep))` would also help here.

```suggestion
        inner_most_candidate = max(candidates, key=len)
        return candidates[inner_most_candidate]
```